File: server/catalog.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path

from .models import (
    Availability,
    CatalogResponse,
    ComparisonKind,
    Competitor,
    CompetitorId,
    Corner,
    MetricDirection,
    MetricRole,
    MetricSpec,
    MetricUnit,
    Persona,
    PresenterLens,
    PresenterPack,
    RedoPresentation,
    RoundDefinition,
    RoundId,
)

ROOT = Path(__file__).resolve().parents[1]

#: The protocol Round 5 runs. The bounded 128-attempt protocol is retired: it measured
#: pooled-path setup and a capped burst, which is not the claim Round 5 makes, and both
#: lanes now hold exactly 10,000 authenticated clients from a shared start.
#:
#: `ROUND5_BOUNDED_PROTOCOL` survives as a name only so that a stored scorecard written
#: under it can still be recognised and labelled as a legacy result rather than silently
#: relabelled with 10,000-client copy it never attempted. It is not selectable.
ROUND5_BOUNDED_PROTOCOL = "connection-spike-v1"
ROUND5_FANIN_PROTOCOL = "round5-fanin-v2"
ROUND5_PROTOCOLS = frozenset({ROUND5_FANIN_PROTOCOL})


COMPETITORS = [
    Competitor(
        id=CompetitorId.AURORA_SERVERLESS_V2,
        name="Amazon Aurora PostgreSQL Serverless v2",
        short_name="Aurora Serverless v2",
        edition="AURORA SERVERLESS v2 EDITION",
    ),
    Competitor(
        id=CompetitorId.RDS_POSTGRES,
        name="Amazon RDS for PostgreSQL",
        short_name="RDS PostgreSQL",
        edition="RDS FOR POSTGRESQL EDITION",
    ),
]


MODEL_SCORE_METRICS = [
    MetricSpec(
        id="managed_availability_ms",
        label="Reverse ETL sync",
        role=MetricRole.PRIMARY,
        unit=MetricUnit.MILLISECONDS,
        direction=MetricDirection.LOWER_IS_BETTER,
    ),
    MetricSpec(
        id="application_proof_elapsed_ms",
        label="End-to-end proof",
        role=MetricRole.SECONDARY,
        unit=MetricUnit.MILLISECONDS,
        direction=MetricDirection.LOWER_IS_BETTER,
    ),
    MetricSpec(
        id="delta_commit_version",
        label="Delta commit version",
        role=MetricRole.GUARDRAIL,
        unit=MetricUnit.VERSION,
        direction=MetricDirection.EXACT,
    ),
    MetricSpec(
        id="exact_row_verified",
        label="Exact row verified",
        role=MetricRole.GUARDRAIL,
        unit=MetricUnit.BOOLEAN,
        direction=MetricDirection.EXACT,
    ),
]


CONNECTION_SPIKE_METRICS = [
    MetricSpec(
        id="setup_elapsed_ms",
        label="Pooled-path setup time",
        role=MetricRole.PRIMARY,
        unit=MetricUnit.MILLISECONDS,
        direction=MetricDirection.LOWER_IS_BETTER,
    ),
    MetricSpec(
        id="successful_clients",
        label="Successful clients",
        role=MetricRole.SECONDARY,
        unit=MetricUnit.COUNT,
        direction=MetricDirection.HIGHER_IS_BETTER,
    ),
    MetricSpec(
        id="application_p99_ms",
        label="Application p99",
        role=MetricRole.SECONDARY,
        unit=MetricUnit.MILLISECONDS,
        direction=MetricDirection.LOWER_IS_BETTER,
    ),
    MetricSpec(
        id="error_clients",
        label="Client errors",
        role=MetricRole.GUARDRAIL,
        unit=MetricUnit.COUNT,
        direction=MetricDirection.LOWER_IS_BETTER,
    ),
]

LIVE_ORDERS_METRICS = [
    MetricSpec(
        id="analytics_available_ms",
        label="Live answer available",
        role=MetricRole.PRIMARY,
        unit=MetricUnit.MILLISECONDS,
        direction=MetricDirection.LOWER_IS_BETTER,
    ),
    MetricSpec(
        id="matching_live_orders",
        label="Exact live orders",
        role=MetricRole.SECONDARY,
        unit=MetricUnit.COUNT,
        direction=MetricDirection.EXACT,
    ),
    MetricSpec(
        id="checkout_verified",
        label="Checkout guardrail",
        role=MetricRole.GUARDRAIL,
        unit=MetricUnit.BOOLEAN,
        direction=MetricDirection.EXACT,
    ),
]


ROUNDS = [
    RoundDefinition(
        id=RoundId.WAKE_IDLE_APP,
        title="Wake this idle app",
        capability="Autoscaling and scale-to-zero",
        scorecard_by_corner={
            Corner.COST: "Published compute and storage rates; billed usage reconciles later",
            Corner.SIMPLICITY: "Automatic wake path and time to a verified transaction",
            Corner.PERFORMANCE: "Eligibility to start at zero, then time to verification",
        },
        competitors=[CompetitorId.AURORA_SERVERLESS_V2, CompetitorId.RDS_POSTGRES],
        availability=Availability.READY,
        redo=RedoPresentation(
            policy="show",
            badge="★ SHOW",
            label="RE-DO ROUND",
            description="Repeat the wake proof to show the same automatic product behavior.",
        ),
    ),
    RoundDefinition(
        id=RoundId.MAKE_SCHEMA_CHANGE_SAFELY,
        title="Make this schema change safely",
        capability="Instant branching and isolated change testing",
        scorecard_by_corner={
            Corner.COST: "Published rates plus developer wait; billed usage reconciles later",
            Corner.SIMPLICITY: "Steps and time to an application-verified isolated change",
            Corner.PERFORMANCE: "Time to create, migrate, and verify an isolated environment",
        },
        competitors=[CompetitorId.RDS_POSTGRES, CompetitorId.AURORA_SERVERLESS_V2],
        availability=Availability.READY,
        redo=RedoPresentation(
            policy="optional",
            badge="OPTIONAL",
            label="RE-DO ROUND",
            description="Repeat only when the room wants another isolated-change proof.",
        ),
    ),
    RoundDefinition(
        id=RoundId.RECOVER_DELETED_ORDER,
        title="Recover this deleted order",
        capability="Point-in-time branching and restore",
        scorecard_by_corner={
            Corner.COST: "Published rates plus recovery wait; billed usage reconciles later",
            Corner.SIMPLICITY: "Steps to the agreed recovery point and verified read",
            Corner.PERFORMANCE: "Verified application RTO at the agreed RPO",
        },
        competitors=[CompetitorId.RDS_POSTGRES, CompetitorId.AURORA_SERVERLESS_V2],
        availability=Availability.READY,
        redo=RedoPresentation(
            policy="skip",
            badge="SKIP",
            label="RE-DO ROUND",
            description=(
                "Hide after success; retain owned recovery cleanup and retry controls "
                "after failure."
            ),
        ),
    ),
    RoundDefinition(
        id=RoundId.PUT_MODEL_SCORE_IN_APP,
        title="Move lakehouse data into live applications",
        capability=(
            "Managed reverse ETL from Unity Catalog Delta to operational Lakebase Postgres"
        ),
        scorecard_by_corner={
            Corner.COST: "Database list rates captured; required reverse ETL remains unpriced",
            Corner.SIMPLICITY: (
                "Analytics Delta to exact operational Postgres application row"
            ),
            Corner.PERFORMANCE: (
                "Reverse ETL sync and end-to-end proof time"
            ),
        },
        competitors=[CompetitorId.RDS_POSTGRES, CompetitorId.AURORA_SERVERLESS_V2],
        availability=Availability.PLANNED,
        metric_specs=MODEL_SCORE_METRICS,
        comparison_kind=ComparisonKind.CAPABILITY_GAP,
        non_claims=[
            (
                "RDS/Aurora are destination databases only; the same outcome requires a "
                "separate reverse-ETL stack that must be selected or built, secured, "
                "networked, configured, monitored, and operated."
            ),
            "The AWS lane was not executed or timed.",
            "No cross-platform speed comparison or margin is claimed.",
            "No dollar savings are claimed.",
            "No eliminated system is claimed.",
            "No full model-serving capability is claimed.",
        ],
        redo=RedoPresentation(
            policy="show",
            badge="★ SHOW",
            label="CHANGE SCORE IN LAKEHOUSE → WATCH APP UPDATE",
            description=(
                "Change this demo's customer risk score from v1 to v2 in the lakehouse, "
                "then watch the same live app record update."
            ),
        ),
    ),
    RoundDefinition(
        id=RoundId.SURVIVE_CONNECTION_SPIKE,
        title="Ready a pooled application path",
        capability="Included pooling compared with a selected AWS managed pooling path",
        scorecard_by_corner={
            Corner.COST: (
                "Published rates include the new RDS Proxy selected for the AWS reference path"
            ),
            Corner.SIMPLICITY: (
                "Included Lakebase pooled endpoint versus 9 journaled mutations for the "
                "selected AWS managed pooling path"
            ),
            Corner.PERFORMANCE: (
                "Primary pooled-path setup time; secondary "
                "128 attempts, maximum 64 concurrent, errors, application p99, "
                "and a separate witness"
            ),
        },
        competitors=[CompetitorId.RDS_POSTGRES, CompetitorId.AURORA_SERVERLESS_V2],
        availability=Availability.PLANNED,
        metric_specs=CONNECTION_SPIKE_METRICS,
        comparison_kind=ComparisonKind.MEASURED,
        non_claims=[
            (
                "From a database-only declared start, Lakebase verifies its included pooled "
                "endpoint with 0 separately provisioned per-bout pooling components and 0 "
                "per-bout pooling infrastructure mutations. Baseline native-login, "
                "ordinary-role, and runner-credential preparation remains outside the clock."
            ),
            (
                "The selected AWS managed pooling path provisions a new RDS Proxy and performs "
                "9 journaled competitor mutations: 1 per-bout Proxy security group, 1 "
                "default-egress change, 4 exact security-group rules, 1 RDS Proxy, 1 "
                "target-group configuration, and 1 target registration. Its setup clock "
                "stops at the exact application transaction."
            ),
            (
                "The IAM service role, runner permission, and dedicated proxy credential "
                "secret or secrets are sealed install-time prerequisites outside the setup "
                "clock. RDS Proxy is the AWS managed pooling option selected for this reference "
                "path, not a universal Aurora or RDS requirement. Direct connections, an "
                "existing RDS Proxy, PgBouncer, and application pooling were not compared."
            ),
            (
                "Phase 2 makes 128 fresh attempts at maximum 64 concurrent, then separately "
                "holds 64 witness clients to verify multiplexing. The phases are sequential "
                "and never overlap."
            ),
            (
                "Application p99 is nearest-rank p99 derived from raw, unrounded "
                "successful-client latencies. Pooled-path setup time is the primary result "
                "and is never added to that secondary p99."
            ),
            (
                "Lakebase's built-in PgBouncer product limit is up to 10,000 client "
                "connections, not PostgreSQL backend sessions or simultaneous transactions. "
                "The recurring bout measures 128 attempts at maximum 64 concurrent, followed "
                "by separate multiplexing proof. Direct AWS connections, an existing Proxy, "
                "sustained throughput, and storm resilience remain outside this comparison."
            ),
            "This is one live proof session, not a benchmark.",
        ],
    ),
    RoundDefinition(
        id=RoundId.ANALYZE_LIVE_ORDERS,
        title="Move live application data into the lakehouse",
        capability="Built-in change feed (CDF) to separate Delta history",
        scorecard_by_corner={
            Corner.COST: "Database list rates captured; required AWS CDC stack remains unpriced",
            Corner.SIMPLICITY: "Built-in change feed: one checkout to one exact Delta answer",
            Corner.PERFORMANCE: (
                "Commit-to-answer freshness; a separate checkout is the correctness guardrail"
            ),
        },
        competitors=[CompetitorId.RDS_POSTGRES, CompetitorId.AURORA_SERVERLESS_V2],
        availability=Availability.PREVIEW,
        metric_specs=LIVE_ORDERS_METRICS,
        comparison_kind=ComparisonKind.CAPABILITY_GAP,
    ),
]


@lru_cache(maxsize=1)
def load_personas() -> tuple[Persona, ...]:
    path = ROOT / "config" / "personas.json"
    raw = json.loads(path.read_text(encoding="utf-8"))
    return tuple(Persona.model_validate(item) for item in raw["personas"])


def catalog(
    model_score_available: bool = False,
    connection_spike_available: bool = False,
    live_orders_available: bool = False,
    round5_protocol: str = ROUND5_FANIN_PROTOCOL,
) -> CatalogResponse:
    return CatalogResponse(
        competitors=COMPETITORS,
        corners=list(Corner),
        personas=list(load_personas()),
        rounds=[
            round_by_id(
                item.id,
                model_score_available=model_score_available,
                connection_spike_available=connection_spike_available,
                live_orders_available=live_orders_available,
                round5_protocol=round5_protocol,
            )
            for item in ROUNDS
        ],
    )


def persona_by_id(persona_id: str) -> Persona:
    try:
        return next(persona for persona in load_personas() if persona.id == persona_id)
    except StopIteration as exc:
        raise ValueError(f"Unknown persona: {persona_id}") from exc


def competitor_by_id(competitor_id: CompetitorId) -> Competitor:
    return next(item for item in COMPETITORS if item.id == competitor_id)


def round_by_id(
    round_id: RoundId,
    model_score_available: bool = False,
    connection_spike_available: bool = False,
    live_orders_available: bool = False,
    round5_protocol: str = ROUND5_FANIN_PROTOCOL,
) -> RoundDefinition:
    if round5_protocol not in ROUND5_PROTOCOLS:
        raise ValueError(f"Unknown Round 5 protocol: {round5_protocol}")
    item = next(item for item in ROUNDS if item.id == round_id)
    if item.id == RoundId.PUT_MODEL_SCORE_IN_APP:
        return item.model_copy(
            update={
                "availability": (
                    Availability.READY if model_score_available else Availability.PLANNED
                )
            },
            deep=True,
        )
    if item.id == RoundId.SURVIVE_CONNECTION_SPIKE:
        return item.model_copy(
            update={
                "availability": (
                    Availability.READY
                    if connection_spike_available
                    else Availability.PLANNED
                )
            },
            deep=True,
        )
    if item.id == RoundId.ANALYZE_LIVE_ORDERS:
        return item.model_copy(
            update={
                "availability": (
                    Availability.READY if live_orders_available else Availability.PREVIEW
                )
            },
            deep=True,
        )
    return item


def recommend_round(
    competitor: CompetitorId,
    primary: Persona,
    model_score_available: bool = False,
    connection_spike_available: bool = False,
    live_orders_available: bool = False,
    round5_protocol: str = ROUND5_FANIN_PROTOCOL,
) -> tuple[RoundDefinition, str]:
    for preferred in primary.recommended_rounds:
        if preferred == "inherit_primary_round":
            continue
        try:
            candidate = round_by_id(
                RoundId(preferred),
                model_score_available=model_score_available,
                connection_spike_available=connection_spike_available,
                live_orders_available=live_orders_available,
                round5_protocol=round5_protocol,
            )
        except ValueError:
            continue
        if competitor in candidate.competitors and candidate.availability == Availability.READY:
            if (
                competitor == CompetitorId.RDS_POSTGRES
                and candidate.id == RoundId.WAKE_IDLE_APP
            ):
                return candidate, (
                    "RDS PostgreSQL has no automatic scale-to-zero wake path; its capability "
                    "is checked before the bell and only Lakebase is timed."
                )
            return candidate, f"Recommended for {primary.role} and executable for this matchup."

    candidate = round_by_id(
        RoundId.WAKE_IDLE_APP,
        model_score_available=model_score_available,
        connection_spike_available=connection_spike_available,
        live_orders_available=live_orders_available,
    )
    if competitor == CompetitorId.RDS_POSTGRES:
        return candidate, (
            "RDS PostgreSQL has no automatic scale-to-zero wake path; its capability is "
            "checked before the bell and only Lakebase is timed."
        )
    return candidate, (
        f"Selected as the strongest honest {competitor_by_id(competitor).short_name} "
        f"round; the {primary.role} lens changes the explanation, not the evidence."
    )


def build_presenter_pack(
    primary: Persona,
    secondary: list[Persona],
    selected_round: RoundDefinition,
    corners: list[Corner],
    competitor: CompetitorId,
) -> PresenterPack:
    def lens(persona: Persona) -> PresenterLens:
        return PresenterLens(
            persona_id=persona.id,
            nickname=persona.nickname,
            role=persona.role,
            interpretation=persona.presenter.interpretation,
            objection=persona.presenter.objection,
            response=persona.presenter.response,
        )

    discovery_question = primary.questions.get("why") or next(iter(primary.questions.values()))
    if len(corners) == 1:
        remembered_metric = selected_round.scorecard_by_corner[corners[0]]
    else:
        measures = {
            Corner.COST: "cost inputs",
            Corner.SIMPLICITY: "workflow simplicity",
            Corner.PERFORMANCE: "elapsed workflow time",
        }
        selected = [measures[corner] for corner in corners]
        metric_list = (
            f"{selected[0]} and {selected[1]}"
            if len(selected) == 2
            else f"{', '.join(selected[:-1])}, and {selected[-1]}"
        )
        remembered_metric = f"{metric_list.capitalize()} to the same verified outcome"
    if selected_round.id == RoundId.PUT_MODEL_SCORE_IN_APP:
        remembered_metric = "Managed Sync exact-version proof and fresh Postgres exact-row read"
    elif selected_round.id == RoundId.SURVIVE_CONNECTION_SPIKE:
        remembered_metric = (
            "Primary pooled-path setup time; secondary 128 attempts at maximum 64 concurrent, "
            "errors, nearest-rank p99, and a separate 64-client witness"
        )
    elif selected_round.id == RoundId.ANALYZE_LIVE_ORDERS:
        remembered_metric = "One exact live order in Delta with checkout still verified"
    if (
        selected_round.id == RoundId.WAKE_IDLE_APP
        and competitor == CompetitorId.RDS_POSTGRES
    ):
        stop_condition = (
            "Lakebase stops after commit + read-back; RDS eligibility is checked before "
            "the bell and not timed."
        )
    elif selected_round.id == RoundId.WAKE_IDLE_APP:
        stop_condition = (
            "Each clock stops after commit and read-back of its run-unique value."
        )
    elif selected_round.id == RoundId.MAKE_SCHEMA_CHANGE_SAFELY:
        stop_condition = (
            "Each clock stops after the identical migration and transaction verify and "
            "the final source check passes."
        )
    elif selected_round.id == RoundId.RECOVER_DELETED_ORDER:
        stop_condition = (
            "Each clock stops after the exact order reads from recovery and remains absent "
            "at the final source check."
        )
    elif selected_round.id == RoundId.PUT_MODEL_SCORE_IN_APP:
        stop_condition = (
            "The clock ends only after the exact committed Delta version is observed synced "
            "and a fresh Postgres read returns the exact row."
        )
    elif selected_round.id == RoundId.SURVIVE_CONNECTION_SPIKE:
        stop_condition = (
            "Lakebase stops after its included pooled endpoint verifies an exact transaction. "
            "The selected AWS managed pooling path stops only after its new RDS Proxy and "
            "exact transaction verify; the 128-attempt maximum-64-concurrent check, separate "
            "64-client witness, and cleanup gates must still pass before any setup winner or "
            "margin is declared."
        )
    elif selected_round.id == RoundId.ANALYZE_LIVE_ORDERS:
        stop_condition = (
            "The clock stops when the exact committed order appears once in Delta history. "
            "The result waits for a separate checkout to commit."
        )
    elif selected_round.availability == Availability.PREVIEW:
        stop_condition = (
            "This preview round is non-executable; it has no verifier or timing boundary."
        )
    else:
        stop_condition = (
            "This planned round is non-executable; it has no verifier or timing boundary."
        )

    return PresenterPack(
        opening=primary.presenter.opening,
        discovery_question=discovery_question,
        risk=primary.presenter.risk,
        stop_condition=stop_condition,
        remembered_metric=remembered_metric,
        primary=lens(primary),
        secondary=[lens(persona) for persona in secondary],
        closing=primary.presenter.closing,
    )

File: frontend/src/catalog.ts
import type {
  CatalogResponse,
  CompetitorId,
  CustomerCorner,
  PersonaDefinition,
  PersonaId,
  RoundDefinition,
  RoundId,
} from './api/types'
import { personaPortraits } from './assets'
import { ROUND_FIVE_DISPLAY_TITLE } from './round5'

export interface LocalRecommendation {
  round_id: RoundId
  reason: string
  metric: string
  presenter_opening: string
}

const personaCopy: Array<[PersonaId, string, string, string, string, RoundId | 'inherit_primary_round']> = [
  ['data_engineer', 'Backfill Bill', 'Data Engineer', 'Another pipeline, just to move one value.', 'How quickly can a new value travel from its source to a working application?', 'put_model_score_in_app'],
  ['software_engineer', 'Stacktrace Jack', 'Software Engineer', 'Needs a test database. Files a ticket. Waits.', 'Let us time the entire path to a safely tested database change.', 'make_schema_change_safely'],
  ['data_analyst', 'Count Query', 'Data Analyst', 'Two dashboards, two numbers, one meeting.', 'A number only matters if it is both correct and current.', 'analyze_live_orders_without_slowing_checkout'],
  ['architect_it', 'Major Pattern', 'Architect / IT', 'Wrote the standard. Watched six teams route around it.', 'The fastest path is useless if teams cannot repeat it safely.', 'make_schema_change_safely'],
  ['data_scientist_ml', 'Doctor Drift', 'Data Scientist / ML', 'The score is fine. Nothing is using it.', 'How quickly does a model output become usable by the application?', 'put_model_score_in_app'],
  ['dba', 'Lockjaw Lucy', 'DBA', 'Owns the restore nobody has rehearsed.', 'Available is not recovered; recovered is when the application reads the right data.', 'recover_deleted_order'],
  ['sre', '3 A.M. Sam', 'SRE', 'Paged for a database that says it is fine.', 'The only status that matters is a verified application transaction.', 'wake_idle_app'],
  ['executive', 'The Big Why', 'Executive', 'Has heard “faster” and wants “so what”.', 'Who benefits, why now, and what one number proves the outcome?', 'inherit_primary_round'],
  ['infosec', 'Cipher Viper', 'Infosec', 'Signs off on things built before anyone asked.', 'A workflow is safe only if its control and evidence survive the path.', 'make_schema_change_safely'],
  ['application_owner', 'Launch-Day Lola', 'Application Owner', 'Launch is Thursday. Checkout is the whole product.', 'Ready means the tested database action works—not that infrastructure says it should.', 'wake_idle_app'],
]

export const PERSONAS: PersonaDefinition[] = personaCopy.map(([id, nickname, role, pain, opening, recommended]) => ({
  id,
  nickname,
  pain,
  role,
  portrait: personaPortraits[id],
  source_status: 'bundled_fallback',
  source_slide: null,
  discovery_order: '',
  recommended_rounds: [recommended],
  questions: {},
  presenter: { opening, risk: '', interpretation: '', objection: '', response: '', closing: '' },
}))

const wakeRound: RoundDefinition = {
  id: 'wake_idle_app',
  title: 'Wake this idle app',
  capability: 'Autoscaling and scale-to-zero',
  scorecard_by_corner: {
    cost: 'Published compute and storage rates; billed usage reconciles later',
    simplicity: 'Automatic wake path and time to a verified transaction',
    performance: 'Eligibility to start at zero, then time to verification',
  },
  competitors: ['aurora_serverless_v2', 'rds_postgres'],
  availability: 'ready',
  redo: {
    policy: 'show', badge: '★ SHOW', label: 'RE-DO ROUND',
    description: 'Repeat the wake proof to show the same automatic product behavior.',
  },
}

const safeChangeRound: RoundDefinition = {
  id: 'make_schema_change_safely',
  title: 'Make this schema change safely',
  capability: 'Instant branching and isolated change testing',
  scorecard_by_corner: {
    cost: 'Published rates plus developer wait; billed usage reconciles later',
    simplicity: 'Steps and time to an application-verified isolated change',
    performance: 'Time to create, migrate, and verify an isolated environment',
  },
  competitors: ['aurora_serverless_v2', 'rds_postgres'],
  availability: 'ready',
  redo: {
    policy: 'optional', badge: 'OPTIONAL', label: 'RE-DO ROUND',
    description: 'Repeat only when the room wants another isolated-change proof.',
  },
}

const recoverRound: RoundDefinition = {
  id: 'recover_deleted_order',
  title: 'Recover this deleted order',
  capability: 'Point-in-time branching and restore',
  scorecard_by_corner: {
    cost: 'Published rates plus recovery wait; billed usage reconciles later',
    simplicity: 'Steps to the agreed recovery point and verified read',
    performance: 'Verified application RTO at the agreed RPO',
  },
  competitors: ['aurora_serverless_v2', 'rds_postgres'],
  availability: 'ready',
  redo: {
    policy: 'skip', badge: 'SKIP', label: 'RE-DO ROUND',
    description: 'Hide after success; retain owned recovery cleanup and retry controls after failure.',
  },
}

const modelScoreRound: RoundDefinition = {
  id: 'put_model_score_in_app',
  title: 'Move lakehouse data into live applications',
  capability: 'Managed reverse ETL from Unity Catalog Delta to operational Lakebase Postgres',
  scorecard_by_corner: {
    cost: 'Database list rates captured; required reverse ETL remains unpriced',
    simplicity: 'Analytics Delta to exact operational Postgres application row',
    performance: 'Reverse ETL sync and end-to-end proof time',
  },
  competitors: ['aurora_serverless_v2', 'rds_postgres'],
  availability: 'planned',
  metric_specs: [
    { id: 'managed_availability_ms', label: 'Reverse ETL sync', role: 'primary', unit: 'milliseconds', direction: 'lower_is_better' },
    { id: 'application_proof_elapsed_ms', label: 'End-to-end proof', role: 'secondary', unit: 'milliseconds', direction: 'lower_is_better' },
    { id: 'delta_commit_version', label: 'Delta commit version', role: 'guardrail', unit: 'version', direction: 'exact' },
    { id: 'exact_row_verified', label: 'Exact row verified', role: 'guardrail', unit: 'boolean', direction: 'exact' },
  ],
  comparison_kind: 'capability_gap',
  non_claims: [
    'RDS/Aurora are destination databases only; the same outcome requires a separate reverse-ETL stack that must be selected or built, secured, networked, configured, monitored, and operated.',
    'The AWS lane was not executed or timed.',
    'No cross-platform speed comparison or margin is claimed.',
    'No dollar savings are claimed.',
    'No eliminated system is claimed.',
    'No full model-serving capability is claimed.',
  ],
  redo: {
    policy: 'show',
    badge: '★ SHOW',
    label: 'CHANGE SCORE IN LAKEHOUSE → WATCH APP UPDATE',
    description: 'Change this demo’s customer risk score from v1 to v2 in the lakehouse, then watch the same live app record update.',
  },
}

const connectionSpikeRound: RoundDefinition = {
  id: 'survive_connection_spike',
  title: ROUND_FIVE_DISPLAY_TITLE,
  capability: 'Built-in connection pooling',
  scorecard_by_corner: {
    cost: 'Published rates include the AWS opponent’s added RDS Proxy minimum',
    simplicity: 'Built-in pooling vs an AWS best-practice RDS Proxy from the declared start',
    performance: 'Readiness setup is primary; the identical connection spike is pass/fail validation',
  },
  competitors: ['aurora_serverless_v2', 'rds_postgres'],
  availability: 'ready',
  metric_specs: [
    { id: 'setup_elapsed_ms', label: 'Setup elapsed', role: 'primary', unit: 'milliseconds', direction: 'lower_is_better' },
    { id: 'successful_clients', label: 'Successful clients', role: 'secondary', unit: 'count', direction: 'higher_is_better' },
    { id: 'application_p99_ms', label: 'Warm-burst application p99', role: 'secondary', unit: 'milliseconds', direction: 'lower_is_better' },
    { id: 'error_clients', label: 'Client errors', role: 'guardrail', unit: 'count', direction: 'lower_is_better' },
  ],
  comparison_kind: 'measured',
  non_claims: [
    'RDS Proxy is AWS best practice for connection spikes. If it is already deployed, its setup delay does not apply. This round compares declared-start readiness, not burst performance.',
    'Lakebase uses its built-in pooled host: 0 separate per-bout pooling components and 0 per-bout pooling infrastructure mutations.',
    'The selected Aurora or RDS lane performs 9 journaled competitor mutations: 1 per-bout Proxy security group, 1 default-egress change, 4 exact security-group rules, 1 RDS Proxy, 1 target-group configuration, and 1 target registration.',
    'IAM service role, runner permission, and dedicated proxy credential secret or secrets are sealed install-time prerequisites outside the setup clock. The AWS design still requires added RDS Proxy, Secrets Manager, IAM, and network configuration.',
    'Warm-burst p99 is secondary validation and is never combined with primary setup elapsed time.',
    'Setup failure or a setup towel produces no winner and no margin.',
    'This is one live proof session, not a benchmark.',
  ],
}

const liveOrdersRound: RoundDefinition = {
  id: 'analyze_live_orders_without_slowing_checkout',
  title: 'Move live application data into the lakehouse',
  capability: 'Built-in change feed (CDF) to separate Delta history',
  scorecard_by_corner: {
    cost: 'Database list rates captured; required AWS CDC stack remains unpriced',
    simplicity: 'Built-in change feed: one checkout to one exact Delta answer',
    performance: 'Commit-to-answer freshness; a separate checkout is the correctness guardrail',
  },
  competitors: ['aurora_serverless_v2', 'rds_postgres'],
  availability: 'preview',
}

export const FALLBACK_CATALOG: CatalogResponse = {
  competitors: [
    { id: 'aurora_serverless_v2', name: 'Amazon Aurora PostgreSQL Serverless v2', short_name: 'Aurora Serverless v2', edition: 'AURORA SERVERLESS v2 EDITION' },
    { id: 'rds_postgres', name: 'Amazon RDS for PostgreSQL', short_name: 'RDS PostgreSQL', edition: 'RDS FOR POSTGRESQL EDITION' },
  ],
  personas: PERSONAS,
  corners: ['cost', 'simplicity', 'performance'],
  rounds: [
    wakeRound,
    safeChangeRound,
    recoverRound,
    modelScoreRound,
    connectionSpikeRound,
    liveOrdersRound,
  ],
}

export function withBundledPortraits(catalog: CatalogResponse): CatalogResponse {
  return {
    ...catalog,
    rounds: catalog.rounds.map((round) => round.id === connectionSpikeRound.id
      ? {
          ...round,
          title: connectionSpikeRound.title,
          scorecard_by_corner: connectionSpikeRound.scorecard_by_corner,
          non_claims: connectionSpikeRound.non_claims,
        }
      : round),
    personas: catalog.personas.map((persona) => ({
      ...persona,
      portrait: personaPortraits[persona.id] ?? persona.portrait,
    })),
  }
}

export function recommend(
  catalog: CatalogResponse,
  competitor: CompetitorId,
  corners: CustomerCorner[],
  personaId: PersonaId,
): LocalRecommendation {
  const persona = catalog.personas.find((candidate) => candidate.id === personaId) ?? catalog.personas[0]
  const preferred = persona.recommended_rounds
    .map((id) => catalog.rounds.find((round) => round.id === id))
    .find((round) => round?.availability === 'ready' && round.competitors.includes(competitor))
  const baselineId: RoundId = 'wake_idle_app'
  const selected = preferred ?? catalog.rounds.find((round) => round.id === baselineId) ?? catalog.rounds[0]
  const reason = competitor === 'rds_postgres' && selected.id === 'wake_idle_app'
    ? 'RDS PostgreSQL has no automatic scale-to-zero wake path; its capability is checked before the bell and only Lakebase is timed.'
    : preferred
    ? `Recommended for ${persona.role} and executable for this matchup.`
    : `Selected as the strongest honest matchup; the ${persona.role} lens changes the explanation, not the evidence.`
  return {
    round_id: selected.id,
    reason,
    metric: metricForCorners(selected, corners),
    presenter_opening: persona.presenter.opening,
  }
}

export function metricForCorners(
  round: RoundDefinition,
  corners: CustomerCorner[],
): string {
  const active = corners.length ? corners : ['performance'] satisfies CustomerCorner[]
  if (active.length === 1) return round.scorecard_by_corner[active[0]]

  const measures: Record<CustomerCorner, string> = {
    cost: 'cost inputs',
    simplicity: 'workflow simplicity',
    performance: 'elapsed workflow time',
  }
  const selected = active.map((corner) => measures[corner])
  const list = selected.length === 2
    ? `${selected[0]} and ${selected[1]}`
    : `${selected.slice(0, -1).join(', ')}, and ${selected.at(-1)}`
  return `${list[0].toUpperCase()}${list.slice(1)} to the same verified outcome`
}

export function stopCondition(roundId: RoundId, competitor?: CompetitorId): string {
  if (roundId === 'wake_idle_app' && competitor === 'rds_postgres') {
    return 'Lakebase stops after commit + read-back; RDS eligibility is checked before the bell and not timed.'
  }
  if (roundId === 'wake_idle_app') {
    return 'Each clock stops after commit and read-back of its run-unique value.'
  }
  if (roundId === 'make_schema_change_safely') {
    return 'Each clock stops after the identical migration and transaction verify and the final source check passes.'
  }
  if (roundId === 'recover_deleted_order') {
    return 'Each clock stops after the exact order reads from recovery and remains absent at the final source check.'
  }
  if (roundId === 'put_model_score_in_app') {
    return 'The clock stops only after the committed Delta version is synchronized and a fresh application connection reads the exact operational Postgres row.'
  }
  if (roundId === 'survive_connection_spike') {
    return 'Each clock stops at a verified pooled application path from the declared start. RDS Proxy is AWS best practice; if it is already deployed, this setup delay does not apply. The identical spike must pass on both lanes.'
  }
  if (roundId === 'analyze_live_orders_without_slowing_checkout') {
    return 'The clock stops when the exact committed order appears once in Delta. The result waits for a separate checkout to commit.'
  }
  return 'This planned round is non-executable; it has no verifier or timing boundary.'
}
